### PR TITLE
Partially support Python 3.10

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

Partially resolve https://github.com/optuna/optuna/issues/2981.

## Description of the changes

- Add Python 3.10 to `python_version ` of CIs files.
- Run tests except for PyTorch-related tests with python 3.10 
    - As tested by https://github.com/optuna/optuna/pull/3179, PyTorch does not support Python 3.10 yet.
- Update `tox.ini`: make integration test consistent with CIs for Python 3.6 and add Python 3.10

### CODE REVIEW

**Comment by nzw0301** [comment]
Can I remove this part?
https://github.com/optuna/optuna/blob/58d8d844264dddf4599cadf08908fb0475a31894/.github/workflows/tests.yml#L51-L54

**Comment by hvy** [comment]
@toshihikoyanase , @keisuke-umezawa could you review this PR? Also, cc. @himkt if you have any inputs.

File: .github/workflows/tests-integration.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ jobs:
 
     strategy:
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9]
+        python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
```
Comment (keisuke-umezawa) [inline .github/workflows/tests-integration.yml:21]: Better to use single quotation to align other lines

File: tests/samplers_tests/test_samplers.py [hunk 2]
```diff
@@ -53,9 +53,15 @@
...
-            lambda: optuna.integration.BoTorchSampler(n_startup_trials=0),
             lambda: optuna.samplers.QMCSampler(),
         ]
+    )
+    # TODO(nzw0301): Remove version constraints if BoTorch supports Python 3.10
+    # or Optuna does not support Python 3.6.
+    + (
+        []
+        if sys.version_info >= (3, 10, 0) or sys.version_info < (3, 7, 0)
+        else [lambda: optuna.integration.BoTorchSampler(n_startup_trials=0)]
```
Comment (keisuke-umezawa) [inline tests/samplers_tests/test_samplers.py:64]: I am not against this solution, but how will it remind us later?

File: tests/samplers_tests/test_samplers.py [hunk 3]
```diff
@@ -53,9 +53,15 @@
         []
         if sys.version_info < (3, 7, 0)
         else [
-            lambda: optuna.integration.BoTorchSampler(n_startup_trials=0),
             lambda: optuna.samplers.QMCSampler(),
         ]
+    )
+    # TODO(nzw0301): Remove version constraints if BoTorch supports Python 3.10
+    # or Optuna does not support Python 3.6.
```
Comment (keisuke-umezawa) [inline tests/samplers_tests/test_samplers.py:60]: I am not against this solution, but how will it remind us later?

**Review by keisuke-umezawa — COMMENTED** [review]
Overall lgtm. I just left a few comments.

File: tests/samplers_tests/test_samplers.py (hunk 2 above)
Comment (nzw0301) [reply tests/samplers_tests/test_samplers.py:64]: As you can see, this PR adds to-do lines to L59 -- L60 in this file. If you mean that the todo is not enough, I would like to add explicit tasks to https://github.com/optuna/optuna/issues/2981 for the followups.

File: .github/workflows/tests-integration.yml (hunk 1 above)
Comment (nzw0301) [reply .github/workflows/tests-integration.yml]: Thank you for your suggestion. I've updated all `python-version` in this PR.

File: tests/samplers_tests/test_samplers.py (hunk 2 above)
Comment (keisuke-umezawa) [reply tests/samplers_tests/test_samplers.py:64]: Thanks! That would be a good solution!

> I would like to add explicit tasks to https://github.com/optuna/optuna/issues/2981 for the followups.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM!

**Comment by hvy** [comment]
Just a comment but if we are to update `setup.py`, perhaps the README badge should be updated as well https://github.com/optuna/optuna/blob/master/README.md?plain=1#L5 . It could very much be a separate PR but I think we have to do it manually?

**Comment by nzw0301** [comment]
Thank you for your comment! Good point. I suppose so. Here is the new link:
https://img.shields.io/badge/python-3.6%20%7C%203.7%20%7C%203.8%20%7C%203.9%20%7C%203.10-blue.

File: .github/workflows/dockerimage.yml [hunk 4]
```diff
@@ -27,13 +27,13 @@ jobs:
...
+    # TODO (nzw0301): Remove `exclude` once integration works with Python 3.10.
     strategy:
       matrix:
-        python_version: ['3.6', '3.7', '3.8', '3.9']
+        python_version: ['3.6', '3.7', '3.8', '3.9', '3.10']
         build_type: ['', 'dev']  # "dev" installs all the dependencies including pytest.
         exclude:
-        - python_version: '3.9'
+        - python_version: '3.10'
           build_type: 'dev'
```
Comment (toshihikoyanase) [inline .github/workflows/dockerimage.yml:37]: Python 3.9-dev image will be newly published. I don't think it problematic because of CI results.

**Review by toshihikoyanase — APPROVED** [review]
I'm sorry for the delayed response. LGTM!

Note that this PR includes the following changes in addition to the Python 3.10 support:
- release of Python 3.9 dev image
- remove Cython from setup of GitHub Actions,

**Pull request merged**